api/services/locations-update.js

```diff
@@ -0,0 +1,250 @@
+      incomingLocation.id = `${incomingLocation.country}-${lastId}`;
+
+      // Add location to batch
+      updatedListOfLocations.push(incomingLocation);
+
+      // Update country counter
+      lastIdInCountry[incomingLocation.country] = lastId;
+    } else {
+      // Location already exist, keep it for future operations
+      updatedListOfLocations.push(
+        Object.assign({}, currentLocation, incomingLocation)
+      );
+    }
+  }
+
+  return sortBy(updatedListOfLocations, ['id']);
+};
+
+export const applyParametersMeta = async function (
+  locations,
+  athenaQueryResults
+) {
+  // Set correct types for coordinates and counts
+  const parametersCounts = athenaQueryResults.map(r => {
+    const { lon, lat } = r;
+    if (typeof lon !== 'undefined') r.lon = parseFloat(lon);
+    if (typeof lat !== 'undefined') r.lat = parseFloat(lat);
+    r.count = parseInt(r.count);
+    return r;
+  });
+
+  return locations.map(l => {
+    // Get parameters for location
+    const parametersCount = filter(parametersCounts, p => {
+      return (
+        l.country === p.country &&
+        ((!l.lon && !p.lon && !l.lat && !p.lat) ||
+          (l.lon === p.lon && l.lat === p.lat))
+      );
+    });
+
+    // Merge parameters into locations, if exist
+    if (parametersCount && parametersCount.length > 0) {
+      l.parameters = parametersCount.map(p => p.parameter).sort();
+      l.countsByMeasurement = parametersCount
+        .map(p => pick(p, ['parameter', 'count']))
+        .sort((a, b) => a.parameter.localeCompare(b.parameter));
+      l.count = l.countsByMeasurement.reduce((acc, i) => acc + i.count, 0);
+    } else {
+      l.parametersCount = 0;
+    }
+
+    return l;
+  });
+};
+
+export const upsertLocations = async function (locations) {
+  const existingIds = await db
+    .select('id')
+    .from('locations')
+    .map(r => r.id);
+
+  const locationsToInsert = [];
+  const locationsToUpdate = [];
+  for (let l of locations) {
+    // Generate WKT from coordinates, if available
+    if (isNumber(l.lon) && isNumber(l.lat)) {
+      l.coordinates = `POINT(${l.lon} ${l.lat})`;
+    }
+
+    // Pick allowed location properties
+    l = pick(l, [
+      'id',
+      'city',
+      'coordinates',
+      'count',
+      'country',
+      'countsByMeasurement',
+      'firstUpdated',
+      'lastUpdated',
+      'lat',
+      'location',
+      'lon',
+      'parameters',
+      'sourceNames',
+      'sourceTypes'
+    ]);
+
+    // Check to insert or update location
+    existingIds.includes(l.id)
+      ? locationsToUpdate.push(l)
+      : locationsToInsert.push(l);
+  }
+
+  await db.transaction(async trx => {
+    // Insert new locations
+    await db.batchInsert('locations', locationsToInsert).transacting(trx);
+
+    // Update existing locations
+    for (const existingLocation of locationsToUpdate) {
+      await trx('locations')
+        .update(existingLocation)
+        .where({ id: existingLocation.id });
+    }
+  });
```

Comment:
Yes, insert operation will be big at the first, then update operation will be the larger one. The update strategy was made that way because current PostgreSQL version (9.4) doesn't support [ON CONFLICT UPDATE](https://www.postgresql.org/docs/9.6/sql-insert.html) operator, which would allow batch upsert. Do you think we can migrate to PostgreSQL 9.6? The impact of moving to 10.x or 11.x it's unclear to me, I wouldn't go for a major release update. Version 9.6 [will be supported until 2021](https://www.postgresql.org/support/versioning/).